### PR TITLE
feat: add etag and collections

### PR DESCRIPTION
**Pull request by PavlenkoV**
This update introduces the addition of ETag to collection responses, allowing clients to track versioning for individual resources within a collection. By including an ETag for each item, we enable efficient caching and conditional operations, reducing unnecessary API calls and improving performance. The ETag is added as a read-only field in the response body to ensure version consistency without altering the API's general object model.

This solution can serve as an interim approach until a standardized HATEOAS implementation is adopted.

### CODE REVIEW

File: standards/collections.md [hunk 1]
```diff
@@ -626,3 +626,54 @@ GET /articles?ordering=title,-reviewRating                  // ordering multiple
...
+  ...
+}
+
+// REQUEST
+GET /resources/123
+
+// RESPONSE
+200 OK
+Content-Type: application/json
+ETag: "33a64df551425fcc55e4d42a148795d9f25f89d4"
```
Comment (omnipitous) [inline standards/collections.md:671]: From our convo this morning the above seems great. This line seems redundant. 
Search returns a list so Header is invalid and inclusion of ETag in Body makes sense (Response Body *not Request body)
Get By ID: There is only 1 so Header works but having the Etag in both the Header and Body feels redundant.. not the worst thing but can we standardize on 1 instead of both?

File: standards/collections.md [hunk 2]
```diff
@@ -626,3 +626,54 @@ GET /articles?ordering=title,-reviewRating                  // ordering multiple
...
+- `ETags` on a collection resource **MAY** exist both for the entire response body as a response header, and as an individual `ETag` property on each resource item in the response body if needed.
+- `ETags` are best represented as metadata in headers, agnostic or separated from the resource model representation. However, some requests to resource collections may benefit from including an `ETag` with reach resource individual resource represented in the collection. Adding an `ETag` to the body of a collection response can help reduce the number of API calls by allowing clients to track changes to individual resources without needing to fetch each resource one by one. The `ETag` will act as a unique identifier for each resource version, ensuring that clients can perform safe and efficient updates or deletions without conflicts with `If-Match` header. If using granular collection `ETags` in the response body, then:
```
Comment (nickclarity) [inline standards/collections.md:636]: Is this a typo?

```suggestion
- `ETags` are best represented as metadata in headers, agnostic or separated from the resource model representation. However, some requests to resource collections may benefit from including an `ETag` with each resource individual resource represented in the collection. Adding an `ETag` to the body of a collection response can help reduce the number of API calls by allowing clients to track changes to individual resources without needing to fetch each resource one by one. The `ETag` will act as a unique identifier for each resource version, ensuring that clients can perform safe and efficient updates or deletions without conflicts with `If-Match` header. If using granular collection `ETags` in the response body, then:
```

Comment (nickclarity) [reply standards/collections.md:636]: > The `ETag` will act as a unique identifier for each resource version

The wording for this is a little confusing to me.  The term "identifier" make me think that an API should expose a way to get a specific resource at a specific version, which I don't think you are trying to say.

Is this comment maybe redundant with the definition in of an etag in [Request Response](request-response.md#etag)?

Comment (travisgosselin) [reply standards/collections.md]: Agreed. Let me rephrase

Comment (travisgosselin) [reply standards/collections.md]: Thoughts on this: ```
`ETags` are best represented as metadata in headers, agnostic or separated from the resource model representation. However, some requests to resource collections may benefit from including an `ETag` with each resource individual resource represented in the collection. Adding an `ETag` to the body of a collection response can help reduce the number of API calls by allowing clients to track changes to individual resources without needing to fetch each resource one by one to make use of request headers like `If-Match`.
```

Comment (nickclarity) [reply standards/collections.md]: I like that better, yes.

File: standards/collections.md [hunk 3]
```diff
@@ -626,3 +626,54 @@ GET /articles?ordering=title,-reviewRating                  // ordering multiple
 GET /articles?orderings=title&orderings=-reviewRating       // never use pluralized "orderings"
 DELETE /articles?ordering=title                             // ordering only applies to GET methods.
 ```
+
+## ETags with Collections
+
+ETags in consideration with resource collections post some additional considerations on what the `ETag` should represent and in what granularity. For more general details about using `ETags` in request and response headers for optimistic concurrency, see the [Request Response](request-response.md#etag) section. When working with ETags in collections, the following considerations should be taken into account:
+
+- `ETags` used in a response header **MUST** represent the entire of the response body, which is the entire collection rather than an individual resource. This is typically less useful in many situations. Working with paged collection results means the response header `ETag` is page-sensitive.
```
Comment (nickclarity) [inline standards/collections.md:634]: What about cases where we have a response that contains a collection of resources, but also "meta" data about the response such as pagination?  Would it be preferable to add an ETag as a sibling of the collection in the body instead of the response header?

Comment (nickclarity) [reply standards/collections.md:634]: Ignore me, I only read it 3 times, lol.

File: standards/collections.md (hunk 1 above)
Comment (travisgosselin) [reply standards/collections.md:671]: To me its similar reasoning to including the status code in the body of error message (impacts logging, convenience, etc). In this case specifically, its the intersection of two patterns, where we want data model consistency between your resources returned in collection or singular, but also consistency to other resources by id that use ETag as the more industry norm pattern for individual results.

I think for consistency, both are necessary depending on how the consumer will make use of it.

**Review by omnipitous — APPROVED** [review]

**Review by nickclarity — APPROVED** [review]

**Comment by travisgosselin** [comment]
@omnipitous / @nickclarity / @PavlenkoV - I appreciate your expertise, effort, and review on this one.
